Automotive crash on start because of null WifiManager

<!-- READ THIS FIRST:
- Make sure you run the latest version of the Android app
- Make sure you run the latest version of Home Assistant
- Make sure to check the Companion docs for troubleshooting and configuration: https://companion.home-assistant.io/
- Make sure the bug you found is not already reported, we love to put work in bugfixes instead of closing duplicate bug reports
  DO NOT DELETE ANY TEXT from this template! All requested information is important.
-->

<!-- If you are reporting an issue for Wear OS please list both devices for the below 3 fields -->
**Home Assistant Android app version(s):** beta-9942-64ee62b8-full

**Android version(s):** ?

**Device model(s):** Android Automotive build for Audi

**Home Assistant version:** n/a

**Last working Home Assistant release (if known):** n/a

**Description of problem, include YAML if issue is related to notifications:** App crashes on launch

<!--
- For Wear OS devices we will need the LogCat logs from the device.
- For Android Auto the logs can be retrieved from the connected device.
- Logs from the device can be taken from Settings > Companion App > Show and Share Logs
-->
**Companion App Logs:**

```
pid: 29196 tid: 29196 AndroidRuntime: FATAL EXCEPTION: main
Process: io.homeassistant.companion.android, PID: 29196
    java.lang.RuntimeException: Error receiving broadcast Intent { act=android.media.RINGER_MODE_CHANGED flg=0x34000010 (has extras) } in io.homeassistant.companion.android.sensors.SensorReceiver@9c0f39a
    at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1689)
    at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
    at android.os.Handler.handleCallback(Handler.java:938)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:201)
    at android.os.Looper.loop(Looper.java:288)
    at android.app.ActivityThread.main(ActivityThread.java:7870)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.NullPointerException
    at io.homeassistant.companion.android.common.data.DataModule$Companion.wifiManager(DataModule.kt:166)
    at io.homeassistant.companion.android.common.data.DataModule_Companion_WifiManagerFactory.wifiManager(DataModule_Companion_WifiManagerFactory.java:44)
    at io.homeassistant.companion.android.DaggerHomeAssistantApplication_HiltComponents_SingletonC$SingletonCImpl$SwitchingProvider.get(DaggerHomeAssistantApplication_HiltComponents_SingletonC.java:1640)
    at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
    at io.homeassistant.companion.android.DaggerHomeAssistantApplication_HiltComponents_SingletonC$SingletonCImpl$SwitchingProvider.get(DaggerHomeAssistantApplication_HiltComponents_SingletonC.java:1634)
    at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
    at io.homeassistant.companion.android.DaggerHomeAssistantApplication_HiltComponents_SingletonC$SingletonCImpl$SwitchingProvider.get(DaggerHomeAssistantApplication_HiltComponents_SingletonC.java:1577)
    at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
    at dagger.internal.DelegateFactory.get(DelegateFactory.java:36)
    at io.homeassistant.companion.android.DaggerHomeAssistantApplication_HiltComponents_SingletonC$SingletonCImpl.injectSensorReceiver2(DaggerHomeAssistantApplication_HiltComponents_SingletonC.java:1509)
    at io.homeassistant.companion.android.DaggerHomeAssistantApplication_HiltComponents_SingletonC$SingletonCImpl.injectSensorReceiver(DaggerHomeAssistantApplication_HiltComponents_SingletonC.java:1430)
    at io.homeassistant.companion.android.sensors.Hilt_SensorReceiver.inject(Hilt_SensorReceiver.java:33)
    at io.homeassistant.companion.android.sensors.Hilt_SensorReceiver.onReceive(Hilt_SensorReceiver.java:25)
    at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1679)
    ... 9 more
```

**Screenshot or video of problem:**

**Additional information:**

 - Device probably doesn't have WiFi, in which case the 'fix' should also hide functionality that depends on this such as internal connection
 - Device does not appear to have GMS but is using full version

--- a/app/src/main/java/io/homeassistant/companion/android/settings/server/ServerSettingsFragment.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/server/ServerSettingsFragment.kt
@@ -122,7 +122,7 @@ class ServerSettingsFragment : ServerSettingsView, PreferenceFragmentCompat() {
         }
 
         findPreference<SwitchPreference>("app_lock_home_bypass")?.let {
-            it.isVisible = findPreference<SwitchPreference>("app_lock")?.isChecked == true
+            it.isVisible = findPreference<SwitchPreference>("app_lock")?.isChecked == true && presenter.hasWifi()
         }
 
         findPreference<EditTextPreference>("session_timeout")?.let { pref ->
@@ -138,6 +138,7 @@ class ServerSettingsFragment : ServerSettingsView, PreferenceFragmentCompat() {
             }
             it.onPreferenceChangeListener =
                 onChangeUrlValidator
+            it.isVisible = presenter.hasWifi()
         }
 
         findPreference<Preference>("connection_external")?.setOnPreferenceClickListener {
@@ -157,6 +158,7 @@ class ServerSettingsFragment : ServerSettingsView, PreferenceFragmentCompat() {
                 onDisplaySsidScreen()
                 return@setOnPreferenceClickListener true
             }
+            it.isVisible = presenter.hasWifi()
         }
 
         findPreference<PreferenceCategory>("security_category")?.isVisible = Build.MODEL != "Quest"
@@ -317,7 +319,7 @@ class ServerSettingsFragment : ServerSettingsView, PreferenceFragmentCompat() {
         // Prevent requesting authentication after just enabling the app lock
         presenter.setAppActive(true)
 
-        findPreference<SwitchPreference>("app_lock_home_bypass")?.isVisible = success
+        findPreference<SwitchPreference>("app_lock_home_bypass")?.isVisible = success && presenter.hasWifi()
         findPreference<EditTextPreference>("session_timeout")?.isVisible = success
         return (result == Authenticator.SUCCESS || result == Authenticator.CANCELED)
     }

--- a/app/src/main/java/io/homeassistant/companion/android/settings/server/ServerSettingsPresenter.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/server/ServerSettingsPresenter.kt
@@ -11,6 +11,7 @@ interface ServerSettingsPresenter {
     fun hasMultipleServers(): Boolean
     fun updateServerName()
     fun updateUrlStatus()
+    fun hasWifi(): Boolean
     fun isSsidUsed(): Boolean
     fun clearSsids()
 

--- a/app/src/main/java/io/homeassistant/companion/android/settings/server/ServerSettingsPresenterImpl.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/server/ServerSettingsPresenterImpl.kt
@@ -3,6 +3,7 @@ package io.homeassistant.companion.android.settings.server
 import android.util.Log
 import androidx.preference.PreferenceDataStore
 import io.homeassistant.companion.android.common.data.servers.ServerManager
+import io.homeassistant.companion.android.common.data.wifi.WifiHelper
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.Job
@@ -12,7 +13,8 @@ import kotlinx.coroutines.runBlocking
 import javax.inject.Inject
 
 class ServerSettingsPresenterImpl @Inject constructor(
-    private val serverManager: ServerManager
+    private val serverManager: ServerManager,
+    private val wifiHelper: WifiHelper
 ) : ServerSettingsPresenter, PreferenceDataStore() {
 
     companion object {
@@ -162,6 +164,8 @@ class ServerSettingsPresenterImpl @Inject constructor(
             view.updateSsids(ssids)
         }
     }
+
+    override fun hasWifi(): Boolean = wifiHelper.hasWifi()
 
     override fun isSsidUsed(): Boolean = runBlocking {
         serverManager.getServer(serverId)?.connection?.internalSsids?.isNotEmpty() == true

--- a/app/src/main/java/io/homeassistant/companion/android/settings/websocket/WebsocketSettingFragment.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/websocket/WebsocketSettingFragment.kt
@@ -22,8 +22,10 @@ import androidx.fragment.app.viewModels
 import com.google.accompanist.themeadapter.material.MdcTheme
 import dagger.hilt.android.AndroidEntryPoint
 import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.common.data.wifi.WifiHelper
 import io.homeassistant.companion.android.settings.SettingViewModel
 import io.homeassistant.companion.android.settings.websocket.views.WebsocketSettingView
+import javax.inject.Inject
 import io.homeassistant.companion.android.common.R as commonR
 
 @AndroidEntryPoint
@@ -32,6 +34,9 @@ class WebsocketSettingFragment : Fragment() {
     companion object {
         const val EXTRA_SERVER = "server"
     }
+
+    @Inject
+    lateinit var wifiHelper: WifiHelper
 
     val viewModel: SettingViewModel by viewModels()
 
@@ -73,6 +78,7 @@ class WebsocketSettingFragment : Fragment() {
                     WebsocketSettingView(
                         websocketSetting = settings.value.websocketSetting,
                         unrestrictedBackgroundAccess = isIgnoringBatteryOptimizations,
+                        hasWifi = wifiHelper.hasWifi(),
                         onSettingChanged = { viewModel.updateWebsocketSetting(serverId, it) },
                         onBackgroundAccessTapped = {
                             if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {

--- a/app/src/main/java/io/homeassistant/companion/android/settings/websocket/views/WebsocketSettingView.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/websocket/views/WebsocketSettingView.kt
@@ -30,6 +30,7 @@ import io.homeassistant.companion.android.util.compose.RadioButtonRow
 fun WebsocketSettingView(
     websocketSetting: WebsocketSetting,
     unrestrictedBackgroundAccess: Boolean,
+    hasWifi: Boolean,
     onSettingChanged: (WebsocketSetting) -> Unit,
     onBackgroundAccessTapped: () -> Unit
 ) {
@@ -55,11 +56,13 @@ fun WebsocketSettingView(
                 selected = websocketSetting == WebsocketSetting.NEVER,
                 onClick = { onSettingChanged(WebsocketSetting.NEVER) }
             )
-            RadioButtonRow(
-                text = stringResource(if (BuildConfig.FLAVOR == "full") R.string.websocket_setting_home_wifi else R.string.websocket_setting_home_wifi_minimal),
-                selected = websocketSetting == WebsocketSetting.HOME_WIFI,
-                onClick = { onSettingChanged(WebsocketSetting.HOME_WIFI) }
-            )
+            if (hasWifi) {
+                RadioButtonRow(
+                    text = stringResource(if (BuildConfig.FLAVOR == "full") R.string.websocket_setting_home_wifi else R.string.websocket_setting_home_wifi_minimal),
+                    selected = websocketSetting == WebsocketSetting.HOME_WIFI,
+                    onClick = { onSettingChanged(WebsocketSetting.HOME_WIFI) }
+                )
+            }
             RadioButtonRow(
                 text = stringResource(if (BuildConfig.FLAVOR == "full") R.string.websocket_setting_while_screen_on else R.string.websocket_setting_while_screen_on_minimal),
                 selected = websocketSetting == WebsocketSetting.SCREEN_ON,

--- a/common/src/main/java/io/homeassistant/companion/android/common/data/DataModule.kt
+++ b/common/src/main/java/io/homeassistant/companion/android/common/data/DataModule.kt
@@ -138,7 +138,7 @@ abstract class DataModule {
 
         @Provides
         @Singleton
-        fun wifiManager(@ApplicationContext appContext: Context) = appContext.getSystemService<WifiManager>()!!
+        fun wifiManager(@ApplicationContext appContext: Context) = appContext.getSystemService<WifiManager>()
     }
 
     @Binds

--- a/common/src/main/java/io/homeassistant/companion/android/common/data/wifi/WifiHelper.kt
+++ b/common/src/main/java/io/homeassistant/companion/android/common/data/wifi/WifiHelper.kt
@@ -7,6 +7,9 @@ interface WifiHelper {
         const val INVALID_BSSID = "02:00:00:00:00:00"
     }
 
+    /** Returns if the device exposes Wi-Fi adapter(s) to apps. To check if Wi-Fi is used, see [isUsingWifi]. */
+    fun hasWifi(): Boolean
+
     /** Returns if the active data connection is using Wi-Fi */
     fun isUsingWifi(): Boolean
 

--- a/common/src/main/java/io/homeassistant/companion/android/common/data/wifi/WifiHelperImpl.kt
+++ b/common/src/main/java/io/homeassistant/companion/android/common/data/wifi/WifiHelperImpl.kt
@@ -9,8 +9,11 @@ import javax.inject.Inject
 @Suppress("DEPRECATION")
 class WifiHelperImpl @Inject constructor(
     private val connectivityManager: ConnectivityManager,
-    private val wifiManager: WifiManager
+    private val wifiManager: WifiManager?
 ) : WifiHelper {
+    override fun hasWifi(): Boolean =
+        wifiManager != null
+
     override fun isUsingWifi(): Boolean =
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
             connectivityManager.activeNetwork?.let {
@@ -42,8 +45,8 @@ class WifiHelperImpl @Inject constructor(
     }
 
     override fun getWifiSsid(): String? =
-        wifiManager.connectionInfo.ssid
+        wifiManager?.connectionInfo?.ssid
 
     override fun getWifiBssid(): String? =
-        wifiManager.connectionInfo.bssid
+        wifiManager?.connectionInfo?.bssid
 }

--- a/common/src/main/java/io/homeassistant/companion/android/common/sensors/NetworkSensorManager.kt
+++ b/common/src/main/java/io/homeassistant/companion/android/common/sensors/NetworkSensorManager.kt
@@ -125,16 +125,20 @@ class NetworkSensorManager : SensorManager {
     override val name: Int
         get() = commonR.string.sensor_name_network
     override suspend fun getAvailableSensors(context: Context): List<SensorManager.BasicSensor> {
-        val list = listOf(
+        val wifiSensors = listOf(
             wifiConnection,
             bssidState,
             wifiIp,
             wifiLinkSpeed,
             wifiState,
             wifiFrequency,
-            wifiSignalStrength,
-            publicIp
+            wifiSignalStrength
         )
+        val list = if (hasWifi(context)) {
+            wifiSensors.plus(publicIp)
+        } else {
+            listOf(publicIp)
+        }
         return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
             list.plus(networkType)
         } else {
@@ -175,8 +179,11 @@ class NetworkSensorManager : SensorManager {
         }
     }
 
+    private fun hasWifi(context: Context): Boolean =
+        context.applicationContext.getSystemService<WifiManager>() != null
+
     private fun updateWifiConnectionSensor(context: Context) {
-        if (!isEnabled(context, wifiConnection)) {
+        if (!isEnabled(context, wifiConnection) || !hasWifi(context)) {
             return
         }
 
@@ -218,7 +225,7 @@ class NetworkSensorManager : SensorManager {
     }
 
     private fun updateBSSIDSensor(context: Context) {
-        if (!isEnabled(context, bssidState)) {
+        if (!isEnabled(context, bssidState) || !hasWifi(context)) {
             return
         }
 
@@ -263,7 +270,7 @@ class NetworkSensorManager : SensorManager {
     }
 
     private fun updateWifiIPSensor(context: Context) {
-        if (!isEnabled(context, wifiIp)) {
+        if (!isEnabled(context, wifiIp) || !hasWifi(context)) {
             return
         }
 
@@ -291,7 +298,7 @@ class NetworkSensorManager : SensorManager {
     }
 
     private fun updateWifiLinkSpeedSensor(context: Context) {
-        if (!isEnabled(context, wifiLinkSpeed)) {
+        if (!isEnabled(context, wifiLinkSpeed) || !hasWifi(context)) {
             return
         }
 
@@ -335,7 +342,7 @@ class NetworkSensorManager : SensorManager {
     }
 
     private fun updateWifiSensor(context: Context) {
-        if (!isEnabled(context, wifiState)) {
+        if (!isEnabled(context, wifiState) || !hasWifi(context)) {
             return
         }
 
@@ -359,7 +366,7 @@ class NetworkSensorManager : SensorManager {
     }
 
     private fun updateWifiFrequencySensor(context: Context) {
-        if (!isEnabled(context, wifiFrequency)) {
+        if (!isEnabled(context, wifiFrequency) || !hasWifi(context)) {
             return
         }
 
@@ -387,7 +394,7 @@ class NetworkSensorManager : SensorManager {
     }
 
     private fun updateWifiSignalStrengthSensor(context: Context) {
-        if (!isEnabled(context, wifiSignalStrength)) {
+        if (!isEnabled(context, wifiSignalStrength) || !hasWifi(context)) {
             return
         }
 

